### ISSUE TITLE
Add support for reading AppSettings as environment variables

### ISSUE THREAD

**Issue by pragnagopa** [Azure/azure-functions-java-worker#146]
Java worker has been updated to use generic worker config https://github.com/Azure/azure-functions-host/commit/bf91f5d6a585428138f273c1787f8c18af0e8eaf

We need a way to read JAVA_OPTS in the function app

**Comment by pragnagopa** [comment]
@TurtleBeach - can you please let me know how do you use JAVA_OPTS?

**Comment by TurtleBeach** [comment]
@pragnagopa The syntax you show above is NOT the syntax that was specified some time ago (guessing January or February). We have been using JAVA_OPTS as the APP SETTING NAME and putting a space-delimited set of values in VALUE. This has been working since we first used starting when someone at MS published the mechanism. Specifically, again, the APP SETTING NAME we have been using is JAVA_OPTS (in the same column in the portal as AzureWebJobsDashboard, e.g.) and specific options like this:
 -DUserStorageName=xxxxx -DUserStorageKey=vvvv -DSysStorageName=yyyy -DSysStorageKey=zzzzz
NOTE that we are NOT using JAVA_OPTS to pass variables to Java - we are using JAVA_OPTS to in effect get environment variables - in the function the call is System.getProperty("UserStorageName");

**Comment by pragnagopa** [comment]
Can you please set  languageWorkers:java:arguments instead for JAVA_OPTS

| APP SETTING NAME  | VALUE |
| ------------- | ------------- |
| languageWorkers:java:arguments   | -DUserStorageName=xxxxx -DUserStorageKey=vvvv -DSysStorageName=yyyy -DSysStorageKey=zzzzz  |

**Comment by TurtleBeach** [comment]
That works. Thank you. But I am still puzzled. Http trigger functions that use the "old" syntax for create response: httpRequest.createResponse(HttpCode, HashMap) are now working (returning what we expect - both for Success (200/201) and error (400, 500) ) with runtime 2.0.11933. Will some future runtime break this?? A few days ago you said we would need to change all httpTrigger function returns to use the ...builder().build() syntax. Is this no longer going to be the case?

**Comment by pragnagopa** [comment]
Yes. Next function runtime release will include breaking changes mentioned here https://github.com/Azure/app-service-announcements/issues/122
I will update the announcement as the release proceeds

**Comment by TurtleBeach** [comment]
Regarding languageWorkers... : (1) can we consider this property to be gospel that will remain valid for future releases, or is this a work-around; and (2) what do we do for the local runtime when we update to 2.0.11933 locally? We currently define JAVA_OPTS as a Windows environment variable, and in addition to the account settings we include the debug setting -agentlib:jdwp=transport....

**Comment by TurtleBeach** [comment]
@pragnagopa Further to the aggravation caused by MS development/release processes:
(1) I finally figured out how to make the debugger work with runtime 2.0.11933, and like most things it is easy once you know. Why can't someone at MS who knows simply state what is necessary (since it is subtly different from the prior mechanism, and relies on the new environment setting that I only know about because I lurk here. Why does MS force everybody to independently discover MS's latest idiosyncrasy?
(2) I do not understand why a released (i.e., running in EAST US) runtime version 2.0.11933 executes code that is on its way to being obsolete, while the development runtime 2.0.11933, and libraries required to build with it, does NOT support the obsolete syntax (I am referring here to createResponse()... vs createResponseBuilder()..
(3) I also do not understand the (in my humble opinion) gratuitous decision to now use an HttpStatus enum as the parameter to createResponseBuilder, rather than standard HTTP protocol integers. At some point the enum MUST be converted back to an integer for use in the HTTP protocol - so what misguided soul decided to interject an enum in the method signature. The HTTP integer values are defined (and correctly named, I might add - INTERNAL_ERROR, not the incorrect INTERNAL_SERVER_ERROR) natively in Java standard libraries. Now Java programmers have to remember naming syntax that is slightly different and in at least one case non standard. It boggles the mind. More to the point, it means that now I am forced to use values that MS has chosen to define. Internally we defined a small number of non-standard values to have meaning that only our system ever sees. Now many of our simple return conventions must be changed, with all the attendant possibilities for interjecting errors - the code calling our functions over HTTP doesn't have access to MS enums.
(4) MOST AGGRAVATING. After painstakingly building, deploying, and testing our function system and seeing that everything ran when deployed on the Azure 2.0.11933 runtime (built on the 2.0.11888 runtime), I foolishly thought I could safely build for the 2.0.11933 runtime (i.e., NO source changes to my working code, other than to add new features). After going hither and yon to get libraries that would let me build a stripped down version of one function, I discovered what I alluded to above - what I build with the 2.0.11933 runtime locally seemingly won't run on Azure since I did in fact have to change the HTTP return syntax (including use of the aforementioned Enum). While luckily the functions deployed all work, I need to add capabilities, which now I cannot do (or at least, I cannot add and deploy, though I can test locally).
THIS PROCESS IS MADDENING - and I think with a little foresight on the part of MS it need not be so aggravating. That's what really ticks me off.

**Comment by TurtleBeach** [comment]
@pragnagopa For environment variables - using the mechanism you specified works both locally and on Azure for Java functions (though locally it is important to add the agent debug library so that Visual Studio Code debugger can connect to the locally running function).
On Azure, I see no way around using an APP SETTING - we don't know a priori what specific machine a function is running on so we can't just set environment variables with "Set name=value" as could be done locally.
In short, your mechanism works (defining an APP SETTING that includes -Dxxx=vvv for each variable. Please make sure that mechanism doesn't get broken.

**Comment by pragnagopa** [comment]
For runtime deployments,
- We are rolling out breaking change since we are preview. Once we GA, our releases will be backward compatible.
- I have updated the release notes with info on how to set debug port for java functions https://github.com/Azure/azure-functions-host/releases/tag/v2.0.11933-alpha. Apologies for missing that earlier as I assumed, VS Code is the default IDE and it has been updated for this release https://github.com/Microsoft/vscode-azurefunctions/commit/094e2f1c3e5d37aedd80c75ab5008a3c75937a61

Runtime: 2.0.11933 does not include Http API breaking changes. You should be able to run and deploy functions without any code changes. If you have issues with current release, please open a separate issue and I will investigate.

**Comment by TurtleBeach** [comment]
@pragnagopa I just re-read your post - you were referring to APP SETTINGS and yes I can read APP SETTINGS and, as I did confirm, your APP SETTING named languageWorkers... does allow setting Environment variables which can be read using Java System.getProperty("settingName");

**Comment by pragnagopa** [comment]
languageWorkers:java:arguments is special AppSetting that functions runtime uses to configure Java.exe
Ideally, you would set following AppSettings

| APP SETTING NAME  | VALUE |
| ------------- | ------------- |
| DUserStorageName   | xxxxx  |   
| DUserStorageKey | vvvv |
| DSysStorageName| yyyy |
|DSysStorageKey | zzzzz |

and your function code will read those properties directly 
```
System.getProperty("DUserStorageName");
....
```
Can you please try this?

**Comment by TurtleBeach** [comment]
Regarding "Runtime 2.0.11933 does not...." I agree that is what I empirically discovered yesterday with deployed functions running on Azure. But then, to get that runtime locally I had to use something related to the CLI that was-beta-32 (sorry but MS naming and locations are all over the place and I can't immediately find the exact reference), and after I downloaded that, the breaking changes did in fact occur - I had to modify source code to use createResponseBuilder()  - which then also led to the discovery that the argument for createResponseBuilder was an enum, not an integer. That is my frustration - two runtimes with the same number - one doesn't break, but in order to build for it, the libraries cause breaking changes. It seems to me that libraries and runtimes are out of synch.

**Comment by TurtleBeach** [comment]
I will try your suggestion as soon as I can deploy a function that was built with the new (breaking) libraries. All of our development environments (2 people on 2 different machines) have now got the libraries that have the breaking changes, and I am not prepared to spend time trying to get back to the old libraries. I hope (maybe pray) that soon you will roll out a new runtime on Azure that (1) has the createResponseBuilder() BUT with integer HTTP code as the argument. Until then we will have to work with functions locally (which access Azure blobs and tables that contain what the functions process).
I will file an issue re: the integer parameter.

**Comment by pragnagopa** [comment]
opened https://github.com/Azure/azure-functions-java-worker/issues/148 for investigating CLI issue with v 2.0.11933

**Comment by TurtleBeach** [comment]
@pragnagopa Belatedly, I have now confirmed that simply defining an APP SETTING in Azure can NOT be read using System.getProperty("settingName") in Java; Also, using System.getProperty("settingName") locally does NOT work. While there may be some other mechanism that works, I can reconfirm that your original suggestion does work (languageWorkers:java:arguments as property name; -Dnnnn=vvvv to set properties (NOTE that -D is "define" followed by the name=value pair). To set multiple properties, each -Dname=value must be separated by a space

**Comment by pragnagopa** [comment]
Thanks for the update. I will look into it.

**Comment by pragnagopa** [comment]
@TurtleBeach - I just tested adding an appSetting and reading it in function code. It works. 
Added [AppSetting ](https://docs.microsoft.com/en-us/azure/azure-functions/functions-how-to-use-azure-function-app-settings) with name myAppSetting and value myAppSettingValue
Sample code
```
context.getLogger().info("myAppSetting "+ System.getenv("myAppSetting"));
```
Logs
```
2018-08-01T20:24:32.291 [Information] Executing 'Functions.HttpTrigger-Java' (Reason='This function was programmatically called via the host APIs.', Id=ae85e84f-be71-4eda-aaa1-887a33d36708)
2018-08-01T20:24:39.249 [Information] myAppSetting myAppSettingValue
2018-08-01T20:24:39.837 [Information] Function "HttpTrigger-Java" (ID: 7ed121a4-6076-4ac8-bb64-0ce4a028cc00) invoked by Java Worker
2018-08-01T20:24:39.966 [Information] Executed 'Functions.HttpTrigger-Java' (Succeeded, Id=ae85e84f-be71-4eda-aaa1-887a33d36708)
```

**Comment by TurtleBeach** [comment]
@pragnagopa This is OK, but do note the difference in the calls. I was (have been) using System.getPROPERTY based on MS recommendations from long ago (sometime early this year). You are using System.getENV. Off the top of my head, I am not sure I care which method I use (I think the only potential issue would be cross platform behavior of getProperty vs getEnv). What I do care to know is that whatever you say works and is supported now will in fact be supported in the future. I am inclined to stay with using languageWorkes... and getProperty() if for no other reason than I currently use that in all my functions. But if the languageWorkers...  APP SETTING is not going to be supported in the future then I will change.

**Comment by pragnagopa** [comment]
languageWorkes:java:arguments is an internal detail for configuring java.exe. I recommend switching to adding  AppSettings and reading via System.getenv(...) as this is an established pattern and I do not anticipate any changes around reading environment variables.
